### PR TITLE
Add upload for raredisease to genotype

### PR DESCRIPTION
**Pull request by rannick**
### Added

- Upload of raredisease cases to genotype

### Changed

- Add _mip suffix to mip related genotype functions

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/upload/gt.py [hunk 1]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
-
-    def _get_samples_sex_mip(self, case_obj: Case, hk_version: Version) -> dict:
-        qc_metrics_file = self.get_qcmetrics_file(hk_version)
-        analysis_sexes = self.analysis_sex(qc_metrics_file)
-        samples_sex = {}
-        for link_obj in case_obj.links:
-            sample_id = link_obj.sample.internal_id
+        return gt_data
+
+    def upload(self, data: dict, replace: bool = False):
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:62]: What is data here?

File: cg/meta/upload/gt.py [hunk 2]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        """Upload data about genotypes for a family of samples."""
+        self.gt.upload(str(data["bcf"]), data["samples_sex"], force=replace)
+
+    @staticmethod
+    def _is_suitable_for_genotype_upload(case: Case) -> bool:
+        """Check if a cancer case is contains WGS and normal sample."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:68]:
```suggestion
        """Check if a cancer case is contains WGS and normal samples."""
```
What is normal here?

File: cg/meta/upload/gt.py [hunk 3]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+
+        samples: list[Sample] = case.samples
+        return any(
+            (not sample.is_tumour and PrepCategory.WHOLE_GENOME_SEQUENCING == sample.prep_category)
+            for sample in samples
+        )
+
+    def _get_genotype_file(self, case_id: str) -> File:
+        tags: set[str] = {GenotypeAnalysisTag.GENOTYPE}
+        hk_genotype = self.hk.get_file_from_latest_version(bundle_name=case_id, tags=tags)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:78]: Improve variable name

Comment (henrikstranneheim) [reply cg/meta/upload/gt.py:78]: or/and add type hint

File: cg/meta/upload/gt.py [hunk 4]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
         return samples_sex
 
-    def _get_samples_sex_balsamic(self, case_obj: Case) -> dict:
-        samples_sex = {}
-        for link_obj in case_obj.links:
-            if link_obj.sample.is_tumour:
-                continue
-            sample_id = link_obj.sample.internal_id
+    def _get_samples_sex_mip_dna(self, case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction (stored Housekeeper QC metrics file)."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:98]:
```suggestion
        """Return sex information from StatusDB and from analysis prediction"""
```
Implementation details should not be part of doc str

File: cg/meta/upload/gt.py [hunk 5]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+    def _get_genotype_file(self, case_id: str) -> File:
+        tags: set[str] = {GenotypeAnalysisTag.GENOTYPE}
+        hk_genotype = self.hk.get_file_from_latest_version(bundle_name=case_id, tags=tags)
+        if not hk_genotype:
+            raise FileNotFoundError(f"Genotype file not found for {case_id}")
+        LOG.debug(f"Found genotype file {hk_genotype.full_path}")
+        return hk_genotype
+
+    def _get_samples_sex_balsamic(self, case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction (UNKNOWN for BALSAMIC)."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:85]:
```suggestion
        """Return sex information from StatusDB and for analysis."""
```

File: cg/meta/upload/gt.py [hunk 6]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
-    def _get_samples_sex_balsamic(self, case_obj: Case) -> dict:
-        samples_sex = {}
-        for link_obj in case_obj.links:
-            if link_obj.sample.is_tumour:
-                continue
-            sample_id = link_obj.sample.internal_id
+    def _get_samples_sex_mip_dna(self, case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction (stored Housekeeper QC metrics file)."""
+        qc_metrics_file: Path = self._get_qcmetrics_file(case_id=case.internal_id)
+        analysis_sexes: dict = self._get_analysis_sex_mip_dna(qc_metrics_file)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:100]:
```suggestion
        analysis_sex: dict = self._get_analysis_sex_mip_dna(qc_metrics_file)
```
A dict is usually singular.

File: cg/meta/upload/gt.py [hunk 7]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
             }
         return samples_sex
 
-    def analysis_sex(self, qc_metrics_file: Path) -> dict:
-        """Fetch analysis sex for each sample of an analysis."""
-        qc_metrics: MIPMetricsDeliverables = self.get_parsed_qc_metrics_data(qc_metrics_file)
+    @staticmethod
+    def _get_analysis_sex_mip_dna(qc_metrics_file: Path) -> dict:
+        """Return analysis sex for each sample of an analysis."""
+        LOG.info("qc_metrics stuff")
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:113]:
```suggestion
```

File: cg/meta/upload/gt.py [hunk 8]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
-        """Fetch a qc_metrics file and return the path"""
-        hk_qcmetrics = self.hk.files(
-            version=hk_version_obj.id, tags=HkMipAnalysisTag.QC_METRICS
-        ).first()
-        LOG.debug(f"Found qc metrics file {hk_qcmetrics.full_path}")
+    def _get_bcf_file(
+        self,
+        case_id: str,
+    ) -> File:
+        """Return a BCF file object. Raises error if nothing is found in the bundle"""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:124]:
```suggestion
        """Return a Housekeeper file object. 
        Raises: FileNotFoundError if nothing is found in the Housekeeper bundle."""
```

File: cg/meta/upload/gt.py [hunk 9]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        LOG.debug("Get genotype files from Housekeeper")
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:125]:
```suggestion
        LOG.debug("Get Genotype files from Housekeeper")
```

File: cg/meta/upload/gt.py [hunk 10]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        genotype_file: File = self._get_genotype_file(case_id=case_id)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:126]:
```suggestion
        genotype_file: File = self._get_genotype_file(case_id)
```

File: cg/meta/upload/gt.py [hunk 11]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        if self._is_variant_file(genotype_file=genotype_file):
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:127]:
```suggestion
        if self._is_variant_file(genotype_file):
```

File: cg/meta/upload/gt.py [hunk 12]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+            LOG.debug(f"Found BCF file {genotype_file.full_path}")
+            return genotype_file
+        raise FileNotFoundError(f"No VCF or BCF file found for the last bundle of {case_id}")
+
+    def _get_qcmetrics_file(self, case_id: str) -> Path:
+        """Return a QC metrics file path."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:133]: Add raise to doc str

File: cg/meta/upload/gt.py [hunk 13]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
-        self.gt.upload(str(data["bcf"]), data["samples_sex"], force=replace)
-
     @staticmethod
-    def _is_variant_file(genotype_file: File):
-        return genotype_file.full_path.endswith("vcf.gz") or genotype_file.full_path.endswith("bcf")
-
-    def _get_genotype_files(self, version_id: int) -> list:
-        return self.hk.files(version=version_id, tags=["genotype"]).all()
+    def _get_samples_sex_raredisease(case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction (stored Housekeeper QC metrics file)."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:151]:
```suggestion
        """Return sex information from StatusDB and from analysis prediction."""
```

File: cg/meta/upload/gt.py [hunk 14]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+                ),
+            }
+        return samples_sex
 
     @staticmethod
-    def is_suitable_for_genotype_upload(case_obj: Case) -> bool:
-        """Check if a cancer case is contains WGS and normal sample."""
+    def _get_analysis_sex_raredisease(qc_metrics_file: Path, sample_id: str) -> str:
+        """Return analysis sex for each sample of an analysis."""
+        print(sample_id)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:167]:
```suggestion
```

File: cg/meta/upload/gt.py [hunk 15]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        qc_metrics: list[MetricsBase] = UploadGenotypesAPI._get_parsed_qc_metrics_data_raredisease(
+            qc_metrics_file
+        )
+        print("all good")
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:171]:
```suggestion
```

File: cg/meta/upload/gt.py [hunk 16]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+        for metric in qc_metrics:
+            print(metric)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:173]:
```suggestion
```

File: cg/meta/upload/gt.py [hunk 17]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+            if metric.name == RAREDISEASE_PREDICTED_SEX_METRIC and metric.id == sample_id:
+                print(metric.name)
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:175]:
```suggestion
```

File: cg/meta/upload/gt.py [hunk 18]
```diff
@@ -43,93 +47,142 @@ def data(self, analysis: Analysis) -> dict:
...
+    def _get_parsed_qc_metrics_data_raredisease(qc_metrics: Path) -> list[MetricsBase]:
+        """Parse and return a QC metrics file."""
+        qcmetrics_raw: dict = ReadFile.get_content_from_file(
+            file_format=FileFormat.YAML, file_path=qc_metrics
         )
+        return [MetricsBase(**metric) for metric in qcmetrics_raw["metrics"]]
+
+    @staticmethod
+    def _is_variant_file(genotype_file: File):
+        return genotype_file.full_path.endswith("vcf.gz") or genotype_file.full_path.endswith("bcf")
```
Comment (henrikstranneheim) [inline cg/meta/upload/gt.py:188]: Can we use constants here?

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 19]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
 
     # THEN assert that the the predicted sex per sample_id is returned
     assert sex == genotype_analysis_sex
 
 
-def test_get_parsed_qc_metrics_data(case_qc_metrics_deliverables: Path):
+def test_get_analysis_sex_raredisease(
+    case_qc_metrics_deliverables_raredisease: Path, genotype_analysis_sex: dict
+):
+    """Test to get the predicted sex from a MIP run using the upload genotypes API"""
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:27]:
```suggestion
    """Test to get the predicted sex from a MIP run using the upload Genotypes API"""
```

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 20]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+    # GIVEN a UploadGenotypesAPI some qcmetrics data
+
+    # WHEN fetching the predicted sex by the analysis
+    sex = UploadGenotypesAPI._get_analysis_sex_raredisease(
+        qc_metrics_file=case_qc_metrics_deliverables_raredisease, sample_id="ADM1"
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:32]: Any fixture for the sample?

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 21]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+    )
+
+    # THEN assert that the the predicted sex per sample_id is returned
+    assert sex == "male"
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:36]: Use constant

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 22]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+
+
+def test_get_parsed_qc_metrics_data_mip(case_qc_metrics_deliverables: Path):
     """Test to get the predicted sex from a MIP run using the upload genotypes API"""
     # GIVEN a UploadGenotypesAPI and the path to a qc_metrics file with case data
 
     # WHEN fetching the predicted sex
-    metrics_object: MIPMetricsDeliverables = UploadGenotypesAPI.get_parsed_qc_metrics_data(
+    metrics_object: MIPMetricsDeliverables = UploadGenotypesAPI._get_parsed_qc_metrics_data_mip_dna(
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:44]: Could we exchange data with deliverables in the method name?

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 23]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
     )
-
     # THEN assert that it was successfully created
     assert isinstance(metrics_object, MIPMetricsDeliverables)
 
 
-def test_get_bcf_file(upload_genotypes_api: UploadGenotypesAPI, case_id: str, timestamp: datetime):
+def test_get_parsed_qc_metrics_data_raredisease(case_qc_metrics_deliverables: Path):
+    """Test to get the predicted sex from a MIP run using the upload genotypes API"""
+    # GIVEN a UploadGenotypesAPI and the path to a qc_metrics file with case data
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:53]:
```suggestion
    # GIVEN a UploadGenotypesAPI and the path to a QC metrics file with case data
```

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 24]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+
+    # WHEN fetching the predicted sex
+    metrics_object = UploadGenotypesAPI._get_parsed_qc_metrics_data_raredisease(
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:56]:
```suggestion
    metrics: list = UploadGenotypesAPI._get_parsed_qc_metrics_data_raredisease(
```

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 25]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+
+    # Then, check if all items in the list are instances of MetricsBase
+    assert all(isinstance(item, MetricsBase) for item in metrics_object)
+
+
+def test_get_bcf_file(upload_genotypes_api: UploadGenotypesAPI, case_id: str):
     """Test to get the predicted sex from a MIP run using the upload genotypes API"""
     # GIVEN a UploadGenotypesAPI populated with some data in housekeeper
-    hk_version = upload_genotypes_api.hk.version(case_id, timestamp)
+    # hk_version = upload_genotypes_api.hk.version(case_id, timestamp)
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:70]:
```suggestion
```

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 26]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
 
     # WHEN fetching the gbcf file with the api
-    gbcf = upload_genotypes_api.get_bcf_file(hk_version)
+    gbcf = upload_genotypes_api._get_bcf_file(case_id)
 
     # THEN assert that the file has the correct tag
     assert "snv-gbcf" in (tag.name for tag in gbcf.tags)
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:76]: Use constant

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 27]
```diff
@@ -13,40 +13,91 @@ def test_get_analysis_sex(case_qc_metrics_deliverables: Path, genotype_analysis_
...
+    # GIVEN an analysis object with a trio
+
+    # GIVEN analysis sex were generated and could be found
+    mocker.patch.object(UploadGenotypesAPI, "_get_analysis_sex_mip_dna")
+    UploadGenotypesAPI._get_analysis_sex_mip_dna.return_value = genotype_analysis_sex
+
+    # WHEN parsing the data
+    result = upload_genotypes_api.get_genotype_data(analysis=analysis_obj)
+
+    # THEN assert that the result looks like expected
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:96]:
```suggestion
    # THEN assert that the the number of samples sex is set
```

File: tests/meta/upload/test_upload_genotypes_api.py [hunk 28]
```diff
@@ -57,11 +108,11 @@ def test_get_data(
...
-    mocker.patch.object(UploadGenotypesAPI, "analysis_sex")
-    UploadGenotypesAPI.analysis_sex.return_value = genotype_analysis_sex
+    mocker.patch.object(UploadGenotypesAPI, "_get_analysis_sex_raredisease")
+    UploadGenotypesAPI._get_analysis_sex_raredisease.return_value = genotype_analysis_sex
 
     # WHEN parsing the data
-    result = upload_genotypes_api.data(analysis=analysis_obj)
+    result = upload_genotypes_api.get_genotype_data(analysis=analysis_obj)
 
     # THEN assert that the result looks like expected
```
Comment (henrikstranneheim) [inline tests/meta/upload/test_upload_genotypes_api.py:117]: Same here

**Review by henrikstranneheim — COMMENTED** [review]
Nice!

**Comment by rannick** [comment]
Thanks for the review @henrikstranneheim!

File: cg/meta/upload/gt.py [hunk 29]
```diff
@@ -43,93 +47,148 @@ def data(self, analysis: Analysis) -> dict:
...
-        ).first()
-        LOG.debug(f"Found qc metrics file {hk_qcmetrics.full_path}")
+    def _get_bcf_file(
+        self,
+        case_id: str,
+    ) -> File:
+        """Return a Housekeeper file object.
+        Raises: FileNotFoundError if nothing is found in the Housekeeper bundle."""
+        LOG.debug("Get Genotype files from Housekeeper")
+        genotype_file: File = self._get_genotype_file(case_id)
```
Comment (islean) [inline cg/meta/upload/gt.py:131]: Are we sure that we do not need to get all genotype files? What happens if the latest file tagged with genotype ends in .csi or .tbi for example?

File: cg/meta/upload/gt.py [hunk 30]
```diff
@@ -43,93 +47,148 @@ def data(self, analysis: Analysis) -> dict:
...
-    def _is_variant_file(genotype_file: File):
-        return genotype_file.full_path.endswith("vcf.gz") or genotype_file.full_path.endswith("bcf")
-
-    def _get_genotype_files(self, version_id: int) -> list:
-        return self.hk.files(version=version_id, tags=["genotype"]).all()
+    def _get_samples_sex_raredisease(case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction."""
+        qc_metrics_file: Path = UploadGenotypesAPI._get_qcmetrics_file(case_id=case.internal_id)
+        samples_sex: dict[str, dict[str, str]] = {}
+        for case_sample in case.links:
```
Comment (islean) [inline cg/meta/upload/gt.py:162]:
```suggestion
        for sample in case.samples:
```

**Review by islean — COMMENTED** [review]
Looks good in general! Not too familiar with genotype upload conditions but looking at the database it seems as if only fetching the latest file tagged with genotype might be a bit brittle?

File: cg/meta/upload/gt.py [hunk 31]
```diff
@@ -43,93 +47,148 @@ def data(self, analysis: Analysis) -> dict:
...
+
+    def upload(self, genotype_load_config: dict, replace: bool = False):
+        """Upload a genotype config for a case."""
+        self.gt.upload(
+            str(genotype_load_config["bcf"]), genotype_load_config["samples_sex"], force=replace
+        )
+
+    @staticmethod
+    def _is_suitable_for_genotype_upload(case: Case) -> bool:
+        """Check if a cancer case contains normal (non-tumor) or WGS samples."""
```
[the remark this answers is not in the document]
Comment (rannick) [reply cg/meta/upload/gt.py:72]: Indeed the docstring is a wrong description.
How about 
"""Returns True if there are any non-tumor WGS samples in the case."""
?

File: cg/apps/housekeeper/hk.py [hunk 32]
```diff
@@ -187,7 +187,7 @@ def get_files(self, bundle: str, tags: list | None = None, version: int | None =
         return self._store.get_files(bundle_name=bundle, tag_names=tags, version_id=version)
 
     def get_latest_file(
-        self, bundle: str, tags: list | None = None, version: int | None = None
+        self, bundle: str | None = None, tags: list | None = None, version: int | None = None
```
[the remark this answers is not in the document]
Comment (rannick) [reply cg/apps/housekeeper/hk.py:190]: Removed

File: cg/meta/upload/gt.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (islean) [reply cg/meta/upload/gt.py:72]: Sounds good to me 👍

File: cg/meta/upload/gt.py (hunk 29 above)
Comment (rannick) [reply cg/meta/upload/gt.py:130]: Raredisease only tags one file with "genotype" and the ending is a must for picking the file up:
https://github.com/Clinical-Genomics/cg/blob/ad73de000e3da39e01b67e5fde48eb2fa865e624/cg/resources/raredisease_bundle_filenames.yaml#L297

Comment (rannick) [reply cg/meta/upload/gt.py:130]: But I can add a check if you think that is necessary

File: cg/meta/upload/gt.py [hunk 33]
```diff
@@ -43,93 +47,148 @@ def data(self, analysis: Analysis) -> dict:
...
             }
         return samples_sex
 
-    def analysis_sex(self, qc_metrics_file: Path) -> dict:
-        """Fetch analysis sex for each sample of an analysis."""
-        qc_metrics: MIPMetricsDeliverables = self.get_parsed_qc_metrics_data(qc_metrics_file)
+    @staticmethod
+    def _get_analysis_sex_mip_dna(qc_metrics_file: Path) -> dict:
+        """Return analysis sex for each sample of an analysis."""
+        LOG.info("qc_metrics stuff")
```
[the remark this answers is not in the document]
Comment (rannick) [reply cg/meta/upload/gt.py]: Hopefully fully removed this time!

File: cg/meta/upload/gt.py (hunk 29 above)
Comment (islean) [reply cg/meta/upload/gt.py:132]: Oh, but is this method not used in `get_genotype_data` which seems to be for all workflows and not only raredisease?

File: cg/meta/upload/gt.py [hunk 34]
```diff
@@ -43,93 +47,163 @@ def data(self, analysis: Analysis) -> dict:
...
+        )
+
+    def remove_index(self, hk_genotypes: list[File]) -> File | None:
+        """
+        Take a list of files and removes all items ending with ".tbi".
+        Returns a single remaining file or raises ValueError if more than one file remains.
+        """
+        filtered_files = [
+            file for file in hk_genotypes if not str(file).endswith(FileExtensions.TBI)
+        ]
```
[the remark this answers is not in the document]
Comment (rannick) [reply cg/meta/upload/gt.py]: Should be done now

File: cg/meta/upload/gt.py [hunk 35]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
+        allowed_extensions = (FileExtensions.BCF, FileExtensions.VCF_GZ)
+        filtered_files = [
+            file for file in hk_genotype_files if file.full_path.endswith(allowed_extensions)
+        ]
+        if len(filtered_files) > 1:
+            raise ValueError(
+                f"Error: Expected only one genotype file, but found {len(filtered_files)} "
+                f"({', '.join(map(str, filtered_files))})."
+            )
+        return filtered_files[0] if filtered_files else None
```
Comment (islean) [inline cg/meta/upload/gt.py:214]:
```suggestion
    def _get_single_genotype_file(hk_genotype_files: list[File]) -> File:
        """
        Returns the single .bcf or .vcf.gz file expected to be found in the provided list. Raises an error if the amount of such files is different from 1.
        """
        allowed_extensions = (FileExtensions.BCF, FileExtensions.VCF_GZ)
        filtered_files = [
            file for file in hk_genotype_files if file.full_path.endswith(allowed_extensions)
        ]
        if len(filtered_files) != 1:
            raise ValueError(
                f"Error: Expecte one genotype file, but found {len(filtered_files)} "
                f"({', '.join(map(str, filtered_files))})."
            )
        return filtered_files[0]
```

Comment (islean) [reply cg/meta/upload/gt.py:214]: Mixed return types are often difficult to work with and it seemed as if there has been an error if no file is found either way.

File: cg/meta/upload/gt.py [hunk 36]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
+        )
+        return [MetricsBase(**metric) for metric in qcmetrics_raw["metrics"]]
 
     @staticmethod
-    def is_suitable_for_genotype_upload(case_obj: Case) -> bool:
-        """Check if a cancer case is contains WGS and normal sample."""
+    def _is_variant_file(genotype_file: File):
+        return genotype_file.full_path.endswith(
+            FileExtensions.VCF_GZ
+        ) or genotype_file.full_path.endswith(FileExtensions.BCF)
```
Comment (islean) [inline cg/meta/upload/gt.py:197]: Is this method needed anymore? Seems as if the filtering on VCF_GZ and BCF has already been done when this is called

File: cg/meta/upload/gt.py [hunk 37]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
-        """Fetch a qc_metrics file and return the path"""
-        hk_qcmetrics = self.hk.files(
-            version=hk_version_obj.id, tags=HkMipAnalysisTag.QC_METRICS
-        ).first()
-        LOG.debug(f"Found qc metrics file {hk_qcmetrics.full_path}")
+    def _get_bcf_file(
+        self,
+        case_id: str,
+    ) -> File:
+        """Return a Housekeeper file object.
```
Comment (islean) [inline cg/meta/upload/gt.py:132]:
```suggestion
```
Should be implied from the type hinted return type

File: cg/meta/upload/gt.py [hunk 38]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
+        Raises: FileNotFoundError if nothing is found in the Housekeeper bundle."""
+        LOG.debug("Get Genotype files from Housekeeper")
+        genotype_file: File = self._get_genotype_file(case_id)
+        if self._is_variant_file(genotype_file):
```
Comment (islean) [inline cg/meta/upload/gt.py:136]: This check already seems to be done in self._get_genotype_file now

File: cg/meta/upload/gt.py [hunk 39]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
-        self.gt.upload(str(data["bcf"]), data["samples_sex"], force=replace)
+    def _get_samples_sex_raredisease(self, case: Case) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction."""
+        qc_metrics_file: Path = self._get_qcmetrics_file(case_id=case.internal_id)
+        samples_sex: dict[str, dict[str, str]] = {}
+        for sample in case.samples:
+            sample_id: str = sample.internal_id
+            samples_sex[sample_id] = {
+                "pedigree": sample.sex,
+                "analysis": UploadGenotypesAPI._get_analysis_sex_raredisease(
```
Comment (islean) [inline cg/meta/upload/gt.py:169]:
```suggestion
                "analysis": self._get_analysis_sex_raredisease(
```
Is there a reason for not calling it via self?

File: cg/meta/upload/gt.py [hunk 40]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
+            }
+        return samples_sex
 
     @staticmethod
-    def _is_variant_file(genotype_file: File):
-        return genotype_file.full_path.endswith("vcf.gz") or genotype_file.full_path.endswith("bcf")
+    def _get_analysis_sex_raredisease(qc_metrics_file: Path, sample_id: str) -> str:
+        """Return analysis sex for each sample of an analysis."""
+        qc_metrics: list[MetricsBase] = (
+            UploadGenotypesAPI._get_parsed_qc_metrics_deliverables_raredisease(qc_metrics_file)
```
Comment (islean) [inline cg/meta/upload/gt.py:179]:
```suggestion
            self._get_parsed_qc_metrics_deliverables_raredisease(qc_metrics_file)
```

File: cg/meta/upload/gt.py [hunk 41]
```diff
@@ -43,93 +47,168 @@ def data(self, analysis: Analysis) -> dict:
...
-        )
+    @staticmethod
+    def _sort_genotype_files(hk_genotype_files: list[File]) -> File | None:
+        """
+        Take a list of files and only keep files finishing with .bcf or .vcf.gz
+        Returns a single remaining file or raises ValueError if more than one file remains.
+        """
+        allowed_extensions = (FileExtensions.BCF, FileExtensions.VCF_GZ)
+        filtered_files = [
+            file for file in hk_genotype_files if file.full_path.endswith(allowed_extensions)
```
Comment (islean) [inline cg/meta/upload/gt.py:207]:
```suggestion
            file for file in hk_genotype_files if self._is_variant_file(file)
```
Or you can use the method here

**Review by islean — APPROVED** [review]
Looks good!

File: cg/meta/upload/gt.py (hunk 38 above)
Comment (rannick) [reply cg/meta/upload/gt.py:136]: Yes. Removed

File: cg/meta/upload/gt.py (hunk 36 above)
Comment (rannick) [reply cg/meta/upload/gt.py]: It is redundant, removed

File: cg/meta/upload/gt.py (hunk 41 above)
Comment (rannick) [reply cg/meta/upload/gt.py:188]: Indeed, is_variant_file now used in this context only

File: cg/meta/upload/gt.py (hunk 40 above)
Comment (rannick) [reply cg/meta/upload/gt.py]: Replaced

File: cg/meta/upload/gt.py (hunk 39 above)
Comment (rannick) [reply cg/meta/upload/gt.py]: No

**Comment by rannick** [comment]
```
Log deploy... done.
cg, version 64.0.1
remote: Enumerating objects: 294, done.
remote: Counting objects: 100% (178/178), done.
remote: Compressing objects: 100% (72/72), done.
remote: Total 96 (delta 76), reused 34 (delta 22), pack-reused 0 (from 0)
Unpacking objects: 100% (96/96), done.
From https://github.com/Clinical-Genomics/cg
   b295134..5f05611  master     -> origin/master
   5a05170..ac5d9d0  3843-same-file-archived-to-pdc-the-same-day -> origin/3843-same-file-archived-to-pdc-the-same-day
   72e6d5b..aaa512c  add-raredisease-scout-upload -> origin/add-raredisease-scout-upload
   3501233..2c6bf67  deliver-sample-file -> origin/deliver-sample-file
 * [new branch]      fix-no-sample-post -> origin/fix-no-sample-post
 * [new branch]      non-nullable-att-pacbio -> origin/non-nullable-att-pacbio
   a26afd3..6cddd76  ordertype-application-creation -> origin/ordertype-application-creation
 * [new tag]         v64.0.1    -> v64.0.1
Already on 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating b295134..5f05611
Fast-forward
 .bumpversion.cfg                                               |   2 +-
 cg/__init__.py                                                 |   2 +-
 cg/cli/upload/genotype.py                                      |   5 +--
 cg/constants/constants.py                                      |   2 +
 cg/constants/housekeeper_tags.py                               |   8 ++++
 cg/constants/nf_analysis.py                                    |   4 +-
 cg/meta/upload/gt.py                                           | 191 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------
 pyproject.toml                                                 |   2 +-
 tests/conftest.py                                              |   6 +++
 tests/fixtures/apps/raredisease/case_metrics_deliverables.yaml | 128 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 tests/meta/upload/balsamic/test_balsamic.py                    |   6 +--
 tests/meta/upload/test_upload_genotypes_api.py                 |  93 ++++++++++++++++++++++++++++++++++--------
 12 files changed, 357 insertions(+), 92 deletions(-)
 create mode 100644 tests/fixtures/apps/raredisease/case_metrics_deliverables.yaml
/home/hiseq.clinical
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
```

**Pull request merged**